server/matchmaker/algorithm/stable_marriage.py

```diff
@@ -0,0 +1,221 @@
+
+from ...decorators import with_logger
+from ..search import Match, Search
+from .matchmaker import Matchmaker, MatchmakingPolicy1v1
+from .random_newbies import RandomlyMatchNewbies
+
+WeightedGraph = Dict[Search, List[Tuple[Search, float]]]
+
+class StableMarriage(MatchmakingPolicy1v1):
+    def find(self, ranks: WeightedGraph) -> Dict[Search, Search]:
+        """Perform the stable matching algorithm until a maximal stable matching
+        is found.
+        """
+        self.matches.clear()
+
+        max_degree = max((len(edges) for edges in ranks.values()), default=0)
+        for i in range(max_degree):
+            self._logger.debug(
+                "Round %i of stable marriage, currently %i matches",
+                i,
+                len(self.matches) // 2,
+            )
+            # Do one round of proposals
+            if len(self.matches) == len(ranks):
+                # Everyone found a match so we are done
+                break
+
+            for search in ranks:
+                if search in self.matches:
+                    continue
+
+                if not ranks[search]:
+                    # Preference list exhausted
+                    continue
+
+                preferred, quality = ranks[search].pop()
+
+                self._logger.debug(
+                    "Quality between %s and %s: %f thresholds: [%f, %f]",
+                    search,
+                    preferred,
+                    quality,
+                    search.match_threshold,
+                    preferred.match_threshold,
+                )
+
+                self._propose(search, preferred, quality)
+
+        return self.matches
+
+    def _propose(self, search: Search, preferred: Search, new_quality: float):
+        """An unmatched search proposes to it's preferred opponent.
+
+        If the opponent is not matched, they become matched. If the opponent is
+        matched, but prefers this new search to its current one, then the opponent
+        unmatches from its previous adversary and matches with the new search instead.
+        """
+        if preferred not in self.matches:
+            self._match(search, preferred)
+            return
+
+        current_match = self.matches[preferred]
+        current_quality = preferred.quality_with(current_match)
+
+        if new_quality > current_quality:
+            # Found a better match
+            self._unmatch(preferred)
+            self._match(search, preferred)
+
+
+@with_logger
+class StableMarriageMatchmaker(Matchmaker):
+    """
+    Runs stable marriage to produce a list of matches
+    and afterwards adds random matchups for previously unmatched new players.
+    """
+    def find(
+        self, searches: Iterable[Search], team_size: int
+    ) -> Tuple[List[Match], List[Search]]:
+        if team_size != 1:
+            self._logger.error(
+                "Invalid team size %i for stable marriage matchmaker will be ignored",
+                team_size,
+            )
```

Comment:
Should there be a return here?

Reply:
No, I wouldn't say so - it will simply ignore what you passed in and the parameter is only there to comply with the `Matchmaker` interface. By "ignore" I mean it will continue to execute as if you had passed a `team_size` of 1 (and implicitly assume that all `Search`es passed in `searches` correspond to a single player).

Reply:
Oh ok. So really the parameter is always ignored, and here it's just logging that something weird is going on. Wouldn't it be better to just `assert team_size == 1` then to avoid programming errors?